# Pandas

**¿Qué es Pandas?**  

Pandas es una biblioteca de código abierto para el lenguaje de programación Python, especializada en el manejo y análisis de datos. Es una herramienta fundamental para cualquier persona que trabaje con conjuntos de datos en Python.

![1](1.png)

![2](2.png)

![3](3.png)

![4](4.png)

![5](5.png)

**Características principales de Pandas:** 

- Estructuras de datos potentes: Define nuevas estructuras de datos como DataFrames y Series, basadas en los arrays de NumPy, pero con funcionalidades más avanzadas para el manejo de datos tabulares y series temporales.  

- Manipulación flexible de datos: Permite leer y escribir datos de diversos formatos comunes, como CSV, Excel, bases de datos SQL y archivos JSON.
- Operaciones de análisis avanzadas: Ofrece una amplia gama de funciones para filtrar, ordenar, agrupar, agregar, combinar y transformar datos de manera eficiente.
- Análisis de series temporales: Brinda herramientas específicas para trabajar con datos de series temporales, como el manejo de fechas, índices de tiempo y 'resampling'.
- Visualización de datos: Integra funciones básicas para la creación de gráficos y visualizaciones de datos.

**Se utiliza para:**

- Cargar y limpiar datos: Importar datos de diversas fuentes, eliminar valores faltantes y corregir errores.
- Manipular y transformar datos: Reordenar, filtrar, agrupar y agregar datos según diferentes criterios.
- Analizar datos: Realizar cálculos estadísticos, identificar patrones y tendencias en los datos.
- Visualizar datos: Crear gráficos y visualizaciones para comunicar los resultados del análisis.

In [2]:
#!pip install numpy

In [3]:
#!pip install pandas

In [4]:
import numpy as np
import pandas as pd

In [5]:
psg_players = pd.Series(['Navas', 'Mbappe', 'Neymar', 'Messi'], index=[1,7,10,30])

In [6]:
psg_players

1      Navas
7     Mbappe
10    Neymar
30     Messi
dtype: object

**Pandas, al crear una `Series` si no se la dan índices los asigna de forma automática:**

In [7]:
ingredientes = pd.Series(['Jamón', 'Aceitunas', 'Pan', 'Queso'])

In [8]:
ingredientes

0        Jamón
1    Aceitunas
2          Pan
3        Queso
dtype: object

### Vamos a usar esta vez diccionarios

In [9]:
dict = {1: 'Navas', 7: 'Mbappe', 10:'Neymar', 30: 'Messi'}
pd.Series(dict)

1      Navas
7     Mbappe
10    Neymar
30     Messi
dtype: object

### ¿Que pasa si quiero definir un array con mas valores?

In [10]:
dict_1 = {'Jugador': ['Navas', 'Mbappe', 'Neymar', 'Messi'],
'Altura':[183.0, 170.0, 185.0, 165.0], 
'Goles': [2, 150,180,200]}

In [11]:
pd.DataFrame(dict_1, index=[1,7,10,30])

Jugador  Altura  Goles
1    Navas   183.0      2
7   Mbappe   170.0    150
10  Neymar   185.0    180
30   Messi   165.0    200

### Sin definir índices

In [12]:
pd.DataFrame(dict_1)

Jugador  Altura  Goles
0   Navas   183.0      2
1  Mbappe   170.0    150
2  Neymar   185.0    180
3   Messi   165.0    200

In [13]:
df_Players = pd.DataFrame(dict_1)

In [14]:
df_Players

Jugador  Altura  Goles
0   Navas   183.0      2
1  Mbappe   170.0    150
2  Neymar   185.0    180
3   Messi   165.0    200

In [15]:
df_Players.columns

Index(['Jugador', 'Altura', 'Goles'], dtype='object')

In [16]:
df_Players.index

RangeIndex(start=0, stop=4, step=1)

In [17]:
df_Players

Jugador  Altura  Goles
0   Navas   183.0      2
1  Mbappe   170.0    150
2  Neymar   185.0    180
3   Messi   165.0    200

### Si tienes datos contenidos en un diccionario de Python, puedes crear una ``` Series``` a partir de ellos pasándole el diccionario:

In [18]:
sdata = {"Ohio": 35000, "Texas": 71000, "Oregon": 16000, "Utah": 5000}
obj3 = pd.Series(sdata)
obj3
    

Ohio      35000
Texas     71000
Oregon    16000
Utah       5000
dtype: int64

### Una `Series`  puede convertirse de nuevo en un diccionario con su método `to_dict` :

In [19]:
obj3.to_dict()

{'Ohio': 35000, 'Texas': 71000, 'Oregon': 16000, 'Utah': 5000}

### Cuando sólo se pasa un diccionario, el índice de la `Series` resultante respetará el orden de las claves según el método `keys` del diccionario, que depende del orden de inserción de las claves.  
### Puede anular esto pasando un índice con las claves del diccionario en el orden en que desea que aparezcan en la `Series` resultante:

In [20]:
sdata = {"Ohio": 35000, "Texas": 71000, "Oregon": 16000, "Utah": 5000}
states = ["California", "Ohio", "Oregon", "Texas"]
obj4 = pd.Series(sdata, index=states)
obj4

California        NaN
Ohio          35000.0
Oregon        16000.0
Texas         71000.0
dtype: float64

### Aquí, tres valores encontrados en sdata se colocaron en los lugares apropiados, pero como no se encontró ningún valor para "California", aparece como `NaN` (Not a Number), que se considera en pandas para marcar valores perdidos o NA. Como "Utah" no se incluyó en estados, se excluye del objeto resultante.
### Utilizaremos los términos "missing", "NA" (Not Available) o "null" indistintamente para referirnos a los datos que faltan. Las funciones `isna` y `notna` de pandas deben utilizarse para detectar datos omitidos:

In [21]:
pd.isna(obj4)

California     True
Ohio          False
Oregon        False
Texas         False
dtype: bool

In [22]:
pd.notna(obj4)

California    False
Ohio           True
Oregon         True
Texas          True
dtype: bool

### `Series` también los tiene como métodos de instancia:

In [23]:
obj4.isna()

California     True
Ohio          False
Oregon        False
Texas         False
dtype: bool

### Mas detalles sobre limpieza y depuración se verá mas adelante.

### Una característica de `Series` útil para muchas aplicaciones es que alinea automáticamente por etiqueta de índice en operaciones aritméticas:

In [24]:
obj3

Ohio      35000
Texas     71000
Oregon    16000
Utah       5000
dtype: int64

In [25]:
obj4

California        NaN
Ohio          35000.0
Oregon        16000.0
Texas         71000.0
dtype: float64

In [26]:
obj3 + obj4

California         NaN
Ohio           70000.0
Oregon         32000.0
Texas         142000.0
Utah               NaN
dtype: float64

### Tanto el propio objeto `Series` como su índice tienen un atributo `name`, que se integra con otras áreas de funcionalidad de pandas:

In [27]:
obj4.name = "population"

In [28]:
obj4.index.name = "state"
obj4

state
California        NaN
Ohio          35000.0
Oregon        16000.0
Texas         71000.0
Name: population, dtype: float64

### El índice de una `Series` puede modificarse "in situ" mediante asignación:

### Definamos el siguiente obj:

In [29]:
obj = pd.Series([4, 7, -5, 3])
obj

0    4
1    7
2   -5
3    3
dtype: int64

### Asignamos "in situ" los indices:

In [30]:
obj.index = ["Bob", "Steve", "Jeff", "Ryan"]
obj

Bob      4
Steve    7
Jeff    -5
Ryan     3
dtype: int64

### DataFrame
Un DataFrame representa una tabla rectangular de datos y contiene una colección ordenada y nombrada de columnas, cada una de las cuales puede ser un tipo de valor diferente (numérico, cadena, booleano, etc.). El DataFrame tiene tanto un índice de fila como de columna; puede considerarse como un diccionario de Series que comparten el mismo índice.
Hay muchas maneras de construir un DataFrame, aunque una de las más comunes es a partir de un diccionario de listas de igual longitud o arrays de NumPy:

In [31]:
data = {"state": ["Ohio", "Ohio", "Ohio", "Nevada", "Nevada", "Nevada"],
        "year": [2000, 2001, 2002, 2001, 2002, 2003],
        "pop": [1.5, 1.7, 3.6, 2.4, 2.9, 3.2]}
frame = pd.DataFrame(data)

El DataFrame resultante tendrá su índice asignado automáticamente, como con Series, y las columnas se colocan según el orden de las claves en los datos (que depende de su orden de inserción en el diccionario):

In [32]:
frame

state  year  pop
0    Ohio  2000  1.5
1    Ohio  2001  1.7
2    Ohio  2002  3.6
3  Nevada  2001  2.4
4  Nevada  2002  2.9
5  Nevada  2003  3.2

Para DataFrames grandes, el método `head` selecciona sólo las cinco primeras filas:

In [33]:
frame.head()

state  year  pop
0    Ohio  2000  1.5
1    Ohio  2001  1.7
2    Ohio  2002  3.6
3  Nevada  2001  2.4
4  Nevada  2002  2.9

Del mismo modo, `tail` devuelve las cinco últimas filas:

In [34]:
frame.tail()

state  year  pop
1    Ohio  2001  1.7
2    Ohio  2002  3.6
3  Nevada  2001  2.4
4  Nevada  2002  2.9
5  Nevada  2003  3.2

Si especifica una secuencia de columnas, las columnas del DataFrame se ordenarán en ese orden:

In [35]:
pd.DataFrame(data, columns=["year", "state", "pop"])

year   state  pop
0  2000    Ohio  1.5
1  2001    Ohio  1.7
2  2002    Ohio  3.6
3  2001  Nevada  2.4
4  2002  Nevada  2.9
5  2003  Nevada  3.2

Si pasa una columna que no está contenida en el diccionario, aparecerá con valores ausentes en el resultado:

In [36]:
frame2 = pd.DataFrame(data, columns=["year", "state", "pop", "debt"])
frame2

year   state  pop debt
0  2000    Ohio  1.5  NaN
1  2001    Ohio  1.7  NaN
2  2002    Ohio  3.6  NaN
3  2001  Nevada  2.4  NaN
4  2002  Nevada  2.9  NaN
5  2003  Nevada  3.2  NaN

In [37]:
frame2.columns

Index(['year', 'state', 'pop', 'debt'], dtype='object')

Una columna de un DataFrame puede recuperarse como una Serie mediante notación tipo diccionario o utilizando la notación de atributo `.` (dot notation):

In [38]:
frame2["state"]

0      Ohio
1      Ohio
2      Ohio
3    Nevada
4    Nevada
5    Nevada
Name: state, dtype: object

In [39]:
frame2.year

0    2000
1    2001
2    2002
3    2001
4    2002
5    2003
Name: year, dtype: int64

`frame2[column]` funciona para cualquier nombre de columna, pero `frame2.column`sólo funciona cuando el nombre de la columna es un nombre de variable Python válido y no entra en conflicto con ninguno de los nombres de método de DataFrame. Por ejemplo, si el nombre de una columna contiene espacios en blanco o símbolos que no sean guiones bajos, no se puede acceder a ella con el método de atributo dot.

Observe que las `Series` devueltas tienen el mismo índice que el DataFrame, y su atributo `name` se ha configurado adecuadamente.

Las filas también pueden recuperarse por posición o nombre con los atributos especiales `iloc` y `loc`.

In [40]:
frame2

year   state  pop debt
0  2000    Ohio  1.5  NaN
1  2001    Ohio  1.7  NaN
2  2002    Ohio  3.6  NaN
3  2001  Nevada  2.4  NaN
4  2002  Nevada  2.9  NaN
5  2003  Nevada  3.2  NaN

In [41]:
frame2.loc[1]

year     2001
state    Ohio
pop       1.7
debt      NaN
Name: 1, dtype: object

In [42]:
frame2.iloc[2]

year     2002
state    Ohio
pop       3.6
debt      NaN
Name: 2, dtype: object

**¿Cuál es la diferencia entre iloc y loc?**

**1- iloc (Index Location):**  
Se utiliza para seleccionar elementos basados en su ubicación (índice de posición).

-- Para las filas, puedes usar números enteros para acceder a las posiciones. Por ejemplo:

- `df.iloc[0]` Selecciona la primera fila.  
- `df.iloc[1]` Selecciona la segunda fila.
- `df.iloc[-1]` Selecciona la última fila.

-- Para las columnas, también puedes usar números enteros:

- `df.iloc[:, 0]`Selecciona la primera columna.  
- `df.iloc[:, 1]`Selecciona la segunda columna.  
- `df.iloc[:, -1]`Selecciona la última columna.

**2- loc (Label Location):**  
Se basa en etiquetas (nombres asignados a filas y columnas). Puedes usar etiquetas personalizadas o nombres de índice para acceder a los datos.

Ejemplos:

- `df.loc[0]` Selecciona la fila con etiqueta 0.
- `df.loc['Etiqueta_o_nombre_fila']` Selecciona la fila con el nombre de la fila o etiqueta.
- `df.loc[:, 'Nombre_Columna']` Selecciona una columna por su nombre.

Vuelta al ejemplo de la celda 42

In [43]:
frame2

year   state  pop debt
0  2000    Ohio  1.5  NaN
1  2001    Ohio  1.7  NaN
2  2002    Ohio  3.6  NaN
3  2001  Nevada  2.4  NaN
4  2002  Nevada  2.9  NaN
5  2003  Nevada  3.2  NaN

Las columnas pueden modificarse por asignación. Por ejemplo, a la columna `debt` vacía se le puede asignar un valor escalar o un array de valores:

In [44]:
frame2["debt"] = 16.5
frame2

year   state  pop  debt
0  2000    Ohio  1.5  16.5
1  2001    Ohio  1.7  16.5
2  2002    Ohio  3.6  16.5
3  2001  Nevada  2.4  16.5
4  2002  Nevada  2.9  16.5
5  2003  Nevada  3.2  16.5

Cuando asigne listas o arrays a una columna, la longitud del valor debe coincidir con la longitud del DataFrame. Si asigna una `Series`, sus etiquetas se realinearán exactamente con el índice del DataFrame, insertando los valores que falten en cualquier valor del índice que no esté presente:

In [45]:
val = pd.Series([-1.2, -1.5, -1.7], index=[2, 4, 5])
frame2["debt"] = val
frame2

year   state  pop  debt
0  2000    Ohio  1.5   NaN
1  2001    Ohio  1.7   NaN
2  2002    Ohio  3.6  -1.2
3  2001  Nevada  2.4   NaN
4  2002  Nevada  2.9  -1.5
5  2003  Nevada  3.2  -1.7

### Al asignar una columna que no existe se creará una columna nueva. La palabra clave `del` borrará columnas como con un diccionario. Como ejemplo, primero se añade una nueva columna de valores booleanos donde la columna `state` es igual a `"Ohio"`:



In [46]:
frame2["eastern"] = frame2["state"] == "Ohio"
frame2

year   state  pop  debt  eastern
0  2000    Ohio  1.5   NaN     True
1  2001    Ohio  1.7   NaN     True
2  2002    Ohio  3.6  -1.2     True
3  2001  Nevada  2.4   NaN    False
4  2002  Nevada  2.9  -1.5    False
5  2003  Nevada  3.2  -1.7    False

### El método `del` se puede utilizar para eliminar esta columna:

In [47]:
del frame2["eastern"]

In [48]:
frame2.columns

Index(['year', 'state', 'pop', 'debt'], dtype='object')

In [49]:
frame2

year   state  pop  debt
0  2000    Ohio  1.5   NaN
1  2001    Ohio  1.7   NaN
2  2002    Ohio  3.6  -1.2
3  2001  Nevada  2.4   NaN
4  2002  Nevada  2.9  -1.5
5  2003  Nevada  3.2  -1.7

In [50]:
frame

state  year  pop
0    Ohio  2000  1.5
1    Ohio  2001  1.7
2    Ohio  2002  3.6
3  Nevada  2001  2.4
4  Nevada  2002  2.9
5  Nevada  2003  3.2

### Cuidado.
La columna devuelta al indexar un DataFrame es una vista de los datos subyacentes, no una copia. Por lo tanto, cualquier modificación en la `Series` se reflejará en el DataFrame. La columna puede copiarse explícitamente con el método `copy` de la `Series`.

Otra forma común de datos es un diccionario anidado de diccionarios:

In [51]:
populations = {"Ohio": {2000: 1.5, 2001: 1.7, 2002: 3.6}, "Nevada": {2001: 2.4, 2002: 2.9}}

Si el diccionario anidado se pasa al DataFrame, pandas interpretará las claves externas del diccionario como las columnas, y las claves internas como los índices de fila:

In [52]:
frame3 = pd.DataFrame(populations)
frame3

Ohio  Nevada
2000   1.5     NaN
2001   1.7     2.4
2002   3.6     2.9

Puede transponer el DataFrame (intercambiar filas y columnas) con una sintaxis similar a la de un array NumPy:

In [53]:
frame3.T

2000  2001  2002
Ohio     1.5   1.7   3.6
Nevada   NaN   2.4   2.9

In [54]:
frame3

Ohio  Nevada
2000   1.5     NaN
2001   1.7     2.4
2002   3.6     2.9

### Cuidado
Tenga en cuenta que la transposición descarta los tipos de datos de columna si las columnas no tienen todas el mismo tipo de datos, por lo que transponer y luego volver a transponer puede perder la información del tipo anterior. En este caso, las columnas se convierten en arrays de objetos Python puros.

Las claves de los diccionarios internos se combinan para formar el índice del resultado. Esto no es cierto si se especifica un índice explícito:

In [55]:
pd.DataFrame(populations, index=[2001, 2002, 2003])

Ohio  Nevada
2001   1.7     2.4
2002   3.6     2.9
2003   NaN     NaN

Los diccionarios de `Series` reciben un tratamiento muy similar:

In [56]:
pdata = {"Ohio": frame3["Ohio"][:-1],"Nevada": frame3["Nevada"][:2]}
# frame3["Ohio"][:-1]: Extrae todos los valores de la columna “Ohio”
# excepto el último valor.

# frame3["Nevada"][:2]:Extrae los primeros dos valores de la columna “Nevada”
# en el DataFrame frame3.
pdata

{'Ohio': 2000    1.5
 2001    1.7
 Name: Ohio, dtype: float64,
 'Nevada': 2000    NaN
 2001    2.4
 Name: Nevada, dtype: float64}

In [57]:
pd.DataFrame(pdata)

Ohio  Nevada
2000   1.5     NaN
2001   1.7     2.4

## Posibles entradas de datos para el constructor DataFrame

<img src="tabla_1.png">

Si el índice y las columnas de un DataFrame tienen definidos atributos de nombre, éstos también se mostrarán:

In [58]:
frame3

Ohio  Nevada
2000   1.5     NaN
2001   1.7     2.4
2002   3.6     2.9

In [59]:
frame3.index.name = "year"
frame3.columns.name = "state"
frame3

state  Ohio  Nevada
year               
2000    1.5     NaN
2001    1.7     2.4
2002    3.6     2.9

A diferencia de `Series`, DataFrame no tiene atributo name. El método `to_numpy` de DataFrame devuelve los datos contenidos en el DataFrame como un `ndarray` bidimensional:

In [60]:
frame3.to_numpy()

array([[1.5, nan],
       [1.7, 2.4],
       [3.6, 2.9]])

Si las columnas del DataFrame son de diferentes tipos de datos, el tipo de datos del array devuelto se elegirá para acomodar todas las columnas, veamos lo que contiene frame2:

In [61]:
frame2

year   state  pop  debt
0  2000    Ohio  1.5   NaN
1  2001    Ohio  1.7   NaN
2  2002    Ohio  3.6  -1.2
3  2001  Nevada  2.4   NaN
4  2002  Nevada  2.9  -1.5
5  2003  Nevada  3.2  -1.7

In [62]:
frame2.to_numpy()

array([[2000, 'Ohio', 1.5, nan],
       [2001, 'Ohio', 1.7, nan],
       [2002, 'Ohio', 3.6, -1.2],
       [2001, 'Nevada', 2.4, nan],
       [2002, 'Nevada', 2.9, -1.5],
       [2003, 'Nevada', 3.2, -1.7]], dtype=object)

### Objetos índice

Los objetos `Index` de pandas son responsables de mantener las etiquetas de los ejes (incluyendo los nombres de las columnas de un DataFrame) y otros metadatos (como el nombre o nombres de los ejes). Cualquier array u otra secuencia de etiquetas que se utilice al construir una `Series` o DataFrame se convierte internamente en un Índice:

In [63]:
obj = pd.Series(np.arange(3), index=["a", "b", "c"])
obj

a    0
b    1
c    2
dtype: int64

In [64]:
index = obj.index
index

Index(['a', 'b', 'c'], dtype='object')

In [65]:
index[1:]

Index(['b', 'c'], dtype='object')

Los objetos índice son inmutables y, por tanto, no pueden ser modificados por el usuario:

In [66]:
index[1] = "d"  # TypeError

TypeError: Index does not support mutable operations

La inmutabilidad hace que sea más seguro compartir objetos índice entre estructuras de datos:

In [ ]:
labels = pd.Index(np.arange(3))
labels

In [67]:
obj2 = pd.Series([1.5, -2.5, 0], index=labels)
obj2

NameError: name 'labels' is not defined

In [68]:
obj2.index is labels

NameError: name 'obj2' is not defined

Además de ser similar a un array, un índice también se comporta como un conjunto (set) de tamaño fijo:

In [69]:
frame3

state  Ohio  Nevada
year               
2000    1.5     NaN
2001    1.7     2.4
2002    3.6     2.9

In [70]:
frame3.columns

Index(['Ohio', 'Nevada'], dtype='object', name='state')

In [71]:
"Ohio" in frame3.columns

True

In [72]:
2003 in frame3.index

False

A diferencia de los conjuntos (`set()`) de Python, un índice de pandas puede contener etiquetas duplicadas:

In [73]:
pd.Index(["foo", "foo", "bar", "bar"])

Index(['foo', 'foo', 'bar', 'bar'], dtype='object')

Las selecciones con etiquetas duplicadas tomarán todas las apariciones de esa etiqueta. Cada Índice tiene una serie de métodos y propiedades para la lógica de conjuntos, que responden a otras preguntas habituales sobre los datos que contiene. Algunas de las más útiles se resumen en:

<img src="tabla_2.png">

### Funciones esenciales
Esta sección es una guia a través de la mecánica fundamental de la interacción con los datos contenidos en una Serie o DataFrame. 

### Reindexación (Reindexing)

Un método importante en los objetos pandas es `reindex`, que significa crear un nuevo objeto con los valores reordenados para alinearlos con el nuevo índice. Consideremos un ejemplo:


In [74]:
obj = pd.Series([4.5, 7.2, -5.3, 3.6], index=["d", "b", "a", "c"])
obj

d    4.5
b    7.2
a   -5.3
c    3.6
dtype: float64

Si se llama a `reindex` en esta `Series`, los datos se reordenan de acuerdo con el nuevo índice, introduciendo los valores que faltan si alguno de los valores del índice no estaba ya presente:

In [75]:
obj2 = obj.reindex(["a", "b", "c", "d", "e"])
obj2

a   -5.3
b    7.2
c    3.6
d    4.5
e    NaN
dtype: float64

En el caso de datos ordenados como series temporales, es posible que desee realizar alguna interpolación o relleno de valores al reindexar. Utilizando un método como `ffill`, que rellena los valores hacia delante:

In [76]:
obj3 = pd.Series(["blue", "purple", "yellow"], index=[0, 2, 4])
obj3

0      blue
2    purple
4    yellow
dtype: object

In [77]:
obj3.reindex(np.arange(6), method="ffill")


0      blue
1      blue
2    purple
3    purple
4    yellow
5    yellow
dtype: object

Con DataFrame, `reindex` puede alterar el índice (de filas), las columnas o ambos. Si sólo se le pasa una secuencia, reindexa las filas del resultado:

In [78]:
frame = pd.DataFrame(np.arange(9).reshape((3, 3)),index=["a", "c", "d"],columns=["Ohio", "Texas", "California"])
frame                    

Ohio  Texas  California
a     0      1           2
c     3      4           5
d     6      7           8

In [79]:
frame2 = frame.reindex(index=["a", "b", "c", "d"])
frame2

Ohio  Texas  California
a   0.0    1.0         2.0
b   NaN    NaN         NaN
c   3.0    4.0         5.0
d   6.0    7.0         8.0

The columns can be reindexed with the `columns` keyword:

In [80]:
frame

Ohio  Texas  California
a     0      1           2
c     3      4           5
d     6      7           8

In [81]:
states = ["Texas", "Utah", "California"]
frame.reindex(columns=states)

Texas  Utah  California
a      1   NaN           2
c      4   NaN           5
d      7   NaN           8

Como `"Ohio"` no estaba en los `states`, los datos de esa columna se eliminan del resultado. Otra forma de hacer `reindex` en un eje concreto es pasar las nuevas etiquetas de eje como argumento posicional y, a continuación, especificar el eje que se va a reindexar con la palabra clave `axis`:

In [82]:
frame.reindex(states, axis="columns")

Texas  Utah  California
a      1   NaN           2
c      4   NaN           5
d      7   NaN           8

Consulte la siguiente tabla para obtener más información sobre los argumentos para reindexar.

<img src="tabla_3.png">

En un DataFrame tambien se puede usar `loc` e `iloc`, para reindexar, y muchos usuarios prefieren hacerlo siempre de esta manera. Esto sólo funciona si todas las nuevas etiquetas de índice ya existen en el DataFrame (mientras que reindex insertará los datos que falten para las nuevas etiquetas):

In [83]:
frame

Ohio  Texas  California
a     0      1           2
c     3      4           5
d     6      7           8

In [84]:
frame.loc[["a", "d", "c"], ["California", "Texas"]]

California  Texas
a           2      1
d           8      7
c           5      4

### Eliminar entradas de un eje

Eliminar una o más entradas de un eje es sencillo si ya tienes un array o una lista de índices sin esas entradas, ya que puedes utilizar el método `reindex` o la indexación basada en `.loc`. Como eso puede requerir un poco de lógica  el método `drop` devolverá un nuevo objeto con el valor o valores indicados eliminados de un eje:

In [85]:
obj = pd.Series(np.arange(5.), index=["a", "b", "c", "d", "e"])
obj

a    0.0
b    1.0
c    2.0
d    3.0
e    4.0
dtype: float64

In [86]:
new_obj = obj.drop("c")
new_obj

a    0.0
b    1.0
d    3.0
e    4.0
dtype: float64

In [87]:
obj.drop(["d", "c"])

a    0.0
b    1.0
e    4.0
dtype: float64

Con DataFrame, los valores índice se pueden eliminar de cualquiera de los ejes. Para ilustrar esto, primero crearemos un DataFrame de ejemplo:

In [88]:
data = pd.DataFrame(np.arange(16).reshape((4, 4)),
                    index=["Ohio", "Colorado", "Utah", "New York"],
                    columns=["one", "two", "three", "four"])
data

one  two  three  four
Ohio        0    1      2     3
Colorado    4    5      6     7
Utah        8    9     10    11
New York   12   13     14    15

Si se llama a `drop` con una secuencia de etiquetas, se eliminarán los valores de las etiquetas de fila (eje 0):

In [89]:
data.drop(index=["Colorado", "Ohio"])

one  two  three  four
Utah        8    9     10    11
New York   12   13     14    15

Para eliminar las etiquetas de las columnas, utilice en su lugar la palabra clave `columns`:

In [90]:
data.drop(columns=["two"])

one  three  four
Ohio        0      2     3
Colorado    4      6     7
Utah        8     10    11
New York   12     14    15

También puede eliminar valores de las columnas pasando `axis=1` ( como en NumPy) o `axis="columns"`:

In [91]:
data.drop("two", axis=1)

one  three  four
Ohio        0      2     3
Colorado    4      6     7
Utah        8     10    11
New York   12     14    15

In [92]:
data.drop(["two", "four"], axis="columns")

one  three
Ohio        0      2
Colorado    4      6
Utah        8     10
New York   12     14

### Indexación, selección y filtrado

La indexación de series (`obj[...]`) funciona de forma análoga a la indexación de arrays de NumPy, con la diferencia de que puedes utilizar los valores índice de la serie en lugar de sólo enteros. He aquí algunos ejemplos:

In [93]:
obj = pd.Series(np.arange(4.), index=["a", "b", "c", "d"])
obj

a    0.0
b    1.0
c    2.0
d    3.0
dtype: float64

In [94]:
obj["b"]

np.float64(1.0)

In [95]:
obj[1]

C:\Users\juanj\AppData\Local\Temp\ipykernel_11568\2469632899.py:1: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  obj[1]


np.float64(1.0)

In [96]:
obj[2:4]

c    2.0
d    3.0
dtype: float64

In [97]:
obj[["b", "a", "d"]]

b    1.0
a    0.0
d    3.0
dtype: float64

In [98]:
obj[[1, 3]]

C:\Users\juanj\AppData\Local\Temp\ipykernel_11568\2982346117.py:1: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  obj[[1, 3]]


b    1.0
d    3.0
dtype: float64

In [99]:
obj[obj < 2]

a    0.0
b    1.0
dtype: float64

Aunque puede seleccionar datos por etiqueta de esta manera, la forma preferida de seleccionar valores de índice es con el operador especial `loc`:

In [100]:
obj.loc[["b", "a", "d"]]


b    1.0
a    0.0
d    3.0
dtype: float64

La razón para preferir `loc` es el tratamiento diferente de los enteros cuando se indexa con `[]`. La indexación normal basada en `[]` tratará los enteros como etiquetas si el índice contiene enteros, por lo que el comportamiento difiere dependiendo del tipo de datos del índice. Por ejemplo:

In [101]:
obj1 = pd.Series([1, 2, 3], index=[2, 0, 1])
obj2 = pd.Series([1, 2, 3], index=["a", "b", "c"])

In [102]:
obj1

2    1
0    2
1    3
dtype: int64

In [103]:
obj2

a    1
b    2
c    3
dtype: int64

In [104]:
obj1[[0, 1, 2]]

0    2
1    3
2    1
dtype: int64

In [105]:
obj2[[0, 1, 2]]

C:\Users\juanj\AppData\Local\Temp\ipykernel_11568\2599987575.py:1: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  obj2[[0, 1, 2]]


a    1
b    2
c    3
dtype: int64

Al utilizar `loc`, la expresión `obj.loc[[0, 1, 2]]` fallará cuando el índice no contenga enteros:

In [106]:
obj2.loc[[0, 1]]

KeyError: "None of [Index([0, 1], dtype='int64')] are in the [index]"

Dado que el operador `loc` indexa exclusivamente con etiquetas, existe también un operador `iloc` que indexa exclusivamente con enteros para trabajar de forma consistente tanto si el índice contiene enteros como si no:

In [107]:
obj1.iloc[[0, 1, 2]]

2    1
0    2
1    3
dtype: int64

In [108]:
obj2.iloc[[0, 1, 2]]

a    1
b    2
c    3
dtype: int64

Precaución: También se puede rebanar (slice) con etiquetas, pero funciona de forma diferente al rebanado(slicing) normal de Python, ya que el punto final es inclusivo:

In [109]:
obj2.loc["b":"c"]

b    2
c    3
dtype: int64

La asignación de valores mediante estos métodos modifica la sección correspondiente de la `Series`:

In [110]:
obj2

a    1
b    2
c    3
dtype: int64

In [111]:
obj2.loc["b":"c"] = 5
obj2

a    1
b    5
c    5
dtype: int64

La indexación en un DataFrame recupera una o más columnas, ya sea con un único valor o con una secuencia:

In [112]:
data = pd.DataFrame(np.arange(16).reshape((4, 4)),
                    index=["Ohio", "Colorado", "Utah", "New York"],
                    columns=["one", "two", "three", "four"])
data

one  two  three  four
Ohio        0    1      2     3
Colorado    4    5      6     7
Utah        8    9     10    11
New York   12   13     14    15

In [113]:
data["two"]

Ohio         1
Colorado     5
Utah         9
New York    13
Name: two, dtype: int64

In [114]:
data[["three", "one"]]

three  one
Ohio          2    0
Colorado      6    4
Utah         10    8
New York     14   12

Este tipo de indexación tiene algunos casos especiales. El primero es el corte (slicing) o la selección de datos con una array booleano:

In [115]:
data[:2]
# selecciona las primeras dos filas del DataFrame.

one  two  three  four
Ohio        0    1      2     3
Colorado    4    5      6     7

In [118]:
data[data["three"] > 5]

one  two  three  four
Colorado    4    5      6     7
Utah        8    9     10    11
New York   12   13     14    15

La sintaxis de selección de filas `data[:2]` se proporciona por comodidad. Si se pasa un único elemento o una lista al operador `[]` selecciona columnas.

 Otro caso de uso es la indexación con un DataFrame booleano, como el producido por una comparación escalar. Considere un DataFrame con todos los valores booleanos producidos por comparación con un valor escalar:

In [119]:
data < 5

one    two  three   four
Ohio       True   True   True   True
Colorado   True  False  False  False
Utah      False  False  False  False
New York  False  False  False  False

Podemos utilizar este DataFrame para asignar el valor 0 a cada ubicación con el valor `True`, así:

In [120]:
data[data < 5] = 0
data

one  two  three  four
Ohio        0    0      0     0
Colorado    0    5      6     7
Utah        8    9     10    11
New York   12   13     14    15

### Selección en un DataFrame con `loc` e `iloc`

Al igual que `Series`, los `DataFrame` tiene atributos especiales `loc` e `iloc` para la indexación basada en etiquetas y en enteros, respectivamente. Como un `DataFrame` es bidimensional, puede seleccionar un subconjunto de filas y columnas con notación tipo NumPy utilizando etiquetas de eje (loc) o enteros (iloc).

Como primer ejemplo, vamos a seleccionar una sola fila por etiqueta:

In [121]:
data

one  two  three  four
Ohio        0    0      0     0
Colorado    0    5      6     7
Utah        8    9     10    11
New York   12   13     14    15

In [122]:
data.loc["Colorado"]

one      0
two      5
three    6
four     7
Name: Colorado, dtype: int64

El resultado de seleccionar una sola fila es una `Series` con un índice que contiene las etiquetas de las columnas del `DataFrame`. Para seleccionar múltiples valores, creando un nuevo DataFrame, se le pasa una secuencia de etiquetas:

In [123]:
data.loc[["Colorado", "New York"]]

one  two  three  four
Colorado    0    5      6     7
New York   12   13     14    15

Puede combinar la selección de filas y columnas en `loc` separando las selecciones con una coma:

In [124]:
data.loc["Colorado", ["two", "three"]]

two      5
three    6
Name: Colorado, dtype: int64

A continuación, realizaremos algunas selecciones similares con enteros utilizando `iloc`:

In [125]:
data.iloc[2]

one       8
two       9
three    10
four     11
Name: Utah, dtype: int64

In [126]:
data.iloc[[2, 1]]

one  two  three  four
Utah        8    9     10    11
Colorado    0    5      6     7

In [127]:
data.iloc[2, [3, 0, 1]]

four    11
one      8
two      9
Name: Utah, dtype: int64

In [128]:
data.iloc[[1, 2], [3, 0, 1]]

four  one  two
Colorado     7    0    5
Utah        11    8    9

Ambas funciones de indexación funcionan con trozos (slices) además de con etiquetas individuales o listas de etiquetas:

In [129]:
data

one  two  three  four
Ohio        0    0      0     0
Colorado    0    5      6     7
Utah        8    9     10    11
New York   12   13     14    15

In [130]:
data.loc[:"Utah", "two"]
# Devuelve los valores de la columna "two" 
# para todas las filas desde el inicio 
# hasta la fila con la etiqueta "Utah"

Ohio        0
Colorado    5
Utah        9
Name: two, dtype: int64

In [131]:
data.iloc[:, :3][data.three > 5]
# Muestra las filas que cumplen
# la condición data.three > 5 
# para las columnas 'one', 'two'
# y 'three'
# Para mayor claridad ejecute 
# data.iloc[:, :3] y luego
# data.iloc[:, :3][data.three > 5]

one  two  three
Colorado    0    5      6
Utah        8    9     10
New York   12   13     14

Los arrays Booleanos pueden ser usados con `loc` pero no `iloc`:

Considere lo siguiente:

In [132]:
data.three >= 2

Ohio        False
Colorado     True
Utah         True
New York     True
Name: three, dtype: bool

In [133]:
data.loc[data.three >= 2]

one  two  three  four
Colorado    0    5      6     7
Utah        8    9     10    11
New York   12   13     14    15

Hay muchas formas de seleccionar y reordenar los datos contenidos en un objeto pandas. Para un `DataFrame`, la siguiente tabla proporciona un breve resumen de muchas de ellas. Como se verá más adelante, hay una serie de opciones adicionales para trabajar con índices jerárquicos.

<img src="tabla_4.png">

### Errores en la indexación de números enteros

Trabajar con objetos pandas indexados por enteros puede ser un escollo para los nuevos usuarios ya que funcionan de forma diferente a las estructuras de datos incorporadas en Python como listas y tuplas. Por ejemplo, es posible que no espere que el siguiente código genere un error:

In [134]:
ser = pd.Series(np.arange(3.))

In [135]:
ser

0    0.0
1    1.0
2    2.0
dtype: float64

In [136]:
ser[-1]

KeyError: -1

En este caso, pandas podría "recurrir" a la indexación por enteros, pero es difícil hacer esto en general sin introducir errores sutiles en el código de usuario. Aquí tenemos un índice que contiene 0, 1 y 2, pero pandas no quiere adivinar lo que quiere el usuario (indexación basada en etiquetas o basada en posiciones):

In [ ]:
ser

En cambio, con un índice no entero, no existe tal ambigüedad:

In [137]:
ser2 = pd.Series(np.arange(3.), index=["a", "b", "c"])
ser2[-1]

C:\Users\juanj\AppData\Local\Temp\ipykernel_11568\821879068.py:2: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  ser2[-1]


np.float64(2.0)

Si tienes un índice de eje que contiene enteros, la selección de datos siempre estará orientada a etiquetas. Como se ha dicho anteriormente, si se utiliza `loc` (para etiquetas) o `iloc` (para enteros) se obtendrá exactamente lo que se desea:

In [138]:
 ser.iloc[-1]

np.float64(2.0)

Por otra parte, el corte (slicing) con números enteros siempre está orientado a números enteros:

In [139]:
ser[:2]

0    0.0
1    1.0
dtype: float64

Como consecuencia de estos errores, es mejor preferir siempre la indexación con `loc` e `iloc` para evitar ambigüedades.

### Errores de la indexación encadenada

En la sección anterior vimos cómo se pueden hacer selecciones flexibles en un DataFrame utilizando `loc` e `iloc`. Estos atributos de indexación también se pueden utilizar para modificar objetos DataFrame "in situ", pero hacerlo requiere cierto cuidado.  
Por ejemplo, en el ejemplo DataFrame anterior, podemos asignar a una columna o fila por etiqueta o posición entera:

In [140]:
data

one  two  three  four
Ohio        0    0      0     0
Colorado    0    5      6     7
Utah        8    9     10    11
New York   12   13     14    15

In [141]:
data.loc[:, "one"] = 1
data

one  two  three  four
Ohio        1    0      0     0
Colorado    1    5      6     7
Utah        1    9     10    11
New York    1   13     14    15

In [142]:
data.iloc[2] = 5
data

one  two  three  four
Ohio        1    0      0     0
Colorado    1    5      6     7
Utah        5    5      5     5
New York    1   13     14    15

In [143]:
data.loc[data["four"] > 5] = 3
data

one  two  three  four
Ohio        1    0      0     0
Colorado    3    3      3     3
Utah        5    5      5     5
New York    3    3      3     3

Un error común para los nuevos usuarios de pandas es encadenar selecciones cuando se asignan de esta manera:

In [144]:
data

one  two  three  four
Ohio        1    0      0     0
Colorado    3    3      3     3
Utah        5    5      5     5
New York    3    3      3     3

In [145]:
data.loc[data.three == 5]
# La condición data.three == 5 
# crea una Serie booleana que 
# tiene True para las filas 
# donde el valor en la columna "three" 
# es igual a 5 y False para las demás filas.

one  two  three  four
Utah    5    5      5     5

In [146]:
data.loc[data.three == 5]["three"] = 6

C:\Users\juanj\AppData\Local\Temp\ipykernel_11568\867481848.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data.loc[data.three == 5]["three"] = 6


Dependiendo del contenido de los datos, esto puede imprimir una advertencia especial `SettingWithCopyWarning`, que le advierte de que está intentando modificar un valor temporal (el resultado no vacío de `data.loc[data.three == 5])` en lugar del DataFrame original `data`, que podría ser lo que pretendía. En este caso, `data` no se ha modificado:

In [147]:
data

one  two  three  four
Ohio        1    0      0     0
Colorado    3    3      3     3
Utah        5    5      5     5
New York    3    3      3     3

En estos casos, la solución consiste en reescribir la asignación encadenada para utilizar una única operación `loc`:

In [148]:
data.loc[data.three == 5, "three"] = 6
data

one  two  three  four
Ohio        1    0      0     0
Colorado    3    3      3     3
Utah        5    5      6     5
New York    3    3      3     3

### Aritmética y alineación de datos

### Pandas puede simplificar mucho el trabajo con objetos que tienen índices diferentes. Por ejemplo, al sumar objetos, si algún par de índices no es el mismo, el índice respectivo en el resultado será la unión de los pares de índices. Veamos un ejemplo:

In [149]:
s1 = pd.Series([7.3, -2.5, 3.4, 1.5], index=["a", "c", "d", "e"])
s1

a    7.3
c   -2.5
d    3.4
e    1.5
dtype: float64

In [150]:
s2 = pd.Series([-2.1, 3.6, -1.5, 4, 3.1],
         index=["a", "c", "e", "f", "g"])
s2

a   -2.1
c    3.6
e   -1.5
f    4.0
g    3.1
dtype: float64

In [151]:
 s1 + s2

a    5.2
c    1.1
d    NaN
e    0.0
f    NaN
g    NaN
dtype: float64

### La alineación interna de los datos introduce valores perdidos en las ubicaciones de las etiquetas que no se solapan. Los valores perdidos se propagarán en los cálculos aritméticos posteriores.

### En el caso de un `DataFrame`, la alineación se realiza tanto en las filas como en las columnas:

In [152]:
df1 = pd.DataFrame(np.arange(9.).reshape((3, 3)), columns=list("bcd"),
             index=["Ohio", "Texas", "Colorado"])

df2 = pd.DataFrame(np.arange(12.).reshape((4, 3)), columns=list("bde"),
                   index=["Utah", "Ohio", "Texas", "Oregon"])

In [153]:
df1

b    c    d
Ohio      0.0  1.0  2.0
Texas     3.0  4.0  5.0
Colorado  6.0  7.0  8.0

In [154]:
df2

b     d     e
Utah    0.0   1.0   2.0
Ohio    3.0   4.0   5.0
Texas   6.0   7.0   8.0
Oregon  9.0  10.0  11.0

In [155]:
df1 + df2

b   c     d   e
Colorado  NaN NaN   NaN NaN
Ohio      3.0 NaN   6.0 NaN
Oregon    NaN NaN   NaN NaN
Texas     9.0 NaN  12.0 NaN
Utah      NaN NaN   NaN NaN

### Como las columnas "c" y "e" no se encuentran en ambos objetos DataFrame, aparecen como ausentes en el resultado. Lo mismo ocurre con las filas con etiquetas que no son comunes a ambos objetos.

### Si añade objetos DataFrame sin etiquetas de columna o fila en común, el resultado contendrá todos nulos:

In [156]:
df1 = pd.DataFrame({"A": [1, 2]})
df2 = pd.DataFrame({"B": [3, 4]})
df1

A
0  1
1  2

In [157]:
df2

B
0  3
1  4

In [158]:
df1 + df2

A   B
0 NaN NaN
1 NaN NaN

## Métodos aritméticos con valores de relleno
### En operaciones aritméticas entre objetos indexados de forma diferente, es posible que desee rellenar con un valor especial, como 0, cuando una etiqueta de eje se encuentra en un objeto pero no en el otro. He aquí un ejemplo en el que establecemos un valor particular como NA (nulo) asignándole `np.nan`:

In [159]:
df1 = pd.DataFrame(np.arange(12.).reshape((3, 4)),
                columns=list("abcd"))
df2 = pd.DataFrame(np.arange(20.).reshape((4, 5)),
              columns=list("abcde"))     

In [160]:
df1

a    b     c     d
0  0.0  1.0   2.0   3.0
1  4.0  5.0   6.0   7.0
2  8.0  9.0  10.0  11.0

In [161]:
df2

a     b     c     d     e
0   0.0   1.0   2.0   3.0   4.0
1   5.0   6.0   7.0   8.0   9.0
2  10.0  11.0  12.0  13.0  14.0
3  15.0  16.0  17.0  18.0  19.0

In [162]:
df1 + df2

a     b     c     d   e
0   0.0   2.0   4.0   6.0 NaN
1   9.0  11.0  13.0  15.0 NaN
2  18.0  20.0  22.0  24.0 NaN
3   NaN   NaN   NaN   NaN NaN

Utilizando el método `add` en df1, se pasa `df2` y un argumento a `fill_value`, que sustituye el valor pasado por cualquier valor que falte en la operación:

In [163]:
df1.add(df2, fill_value=0)

a     b     c     d     e
0   0.0   2.0   4.0   6.0   4.0
1   9.0  11.0  13.0  15.0   9.0
2  18.0  20.0  22.0  24.0  14.0
3  15.0  16.0  17.0  18.0  19.0

En la tabla dada mas bajo encontrará un listado de los métodos `Series` y `DataFrame` para aritmética. Cada uno tiene una contrapartida, que empieza por la letra `r`, que tiene los argumentos invertidos. Por lo tanto, estas dos sentencias son equivalentes:

In [164]:
1 / df1

a         b         c         d
0    inf  1.000000  0.500000  0.333333
1  0.250  0.200000  0.166667  0.142857
2  0.125  0.111111  0.100000  0.090909

In [165]:
df1.rdiv(1)

a         b         c         d
0    inf  1.000000  0.500000  0.333333
1  0.250  0.200000  0.166667  0.142857
2  0.125  0.111111  0.100000  0.090909

<img src="tabla_5.png">

Al reindexar una `Serie` o un `DataFrame`, también puede especificar un valor de relleno (fill value) diferente:

In [166]:
df1

a    b     c     d
0  0.0  1.0   2.0   3.0
1  4.0  5.0   6.0   7.0
2  8.0  9.0  10.0  11.0

In [167]:
df2

a     b     c     d     e
0   0.0   1.0   2.0   3.0   4.0
1   5.0   6.0   7.0   8.0   9.0
2  10.0  11.0  12.0  13.0  14.0
3  15.0  16.0  17.0  18.0  19.0

In [168]:
df1.reindex(columns=df2.columns, fill_value=0)
# df1.reindex(columns=): Toma el DataFrame df1 y lo "reindexa" 
# en base a las columnas de otro DataFrame, df2.
# En otras palabras, asegura que df1 tenga las mismas columnas que df2.
####
# df2.columns: Especifica que las columnas de referencia 
# para el reindexado son las de df2. 
# Esto significa que se agregarán, eliminarán o reordenarán
# las columnas de df1 para que coincidan exactamente con las de df2.
####
# fill_value=0: Este argumento indica el valor que se usará para 
# rellenar las celdas que no existían en df1 pero que sí se 
# encuentran en df2. En este caso, se utiliza el valor 0.

a    b     c     d  e
0  0.0  1.0   2.0   3.0  0
1  4.0  5.0   6.0   7.0  0
2  8.0  9.0  10.0  11.0  0

Repetir código anterior cambiando df1 por df2 y donde aparezca df2 lo cambia por df1. Explicar la salida.

### Operaciones entre `DataFrame` y `Series`
Al igual que con las matrices NumPy de diferentes dimensiones, también se define la aritmética entre DataFrame y Series. En primer lugar, como ejemplo , considere la diferencia entre una matriz bidimensional y una de sus filas:

In [169]:
arr = np.arange(12.).reshape((3, 4))
arr

array([[ 0.,  1.,  2.,  3.],
       [ 4.,  5.,  6.,  7.],
       [ 8.,  9., 10., 11.]])

In [170]:
arr[0]

array([0., 1., 2., 3.])

In [171]:
arr - arr[0]

array([[0., 0., 0., 0.],
       [4., 4., 4., 4.],
       [8., 8., 8., 8.]])

Cuando restamos `arr[0]` de `arr`, la resta se realiza una vez por cada fila. Esto se denomina difusión y se explica con más detalle en lo que se refiere a las arrays generales de NumPy Avanzado. Las operaciones entre un DataFrame y una Serie son similares:

In [172]:
frame = pd.DataFrame(np.arange(12.).reshape((4, 3)),
                    columns=list("bde"),
                    index=["Utah", "Ohio", "Texas", "Oregon"])
series = frame.iloc[0]

In [173]:
frame

b     d     e
Utah    0.0   1.0   2.0
Ohio    3.0   4.0   5.0
Texas   6.0   7.0   8.0
Oregon  9.0  10.0  11.0

In [174]:
series

b    0.0
d    1.0
e    2.0
Name: Utah, dtype: float64

Por defecto, la aritmética entre el `DataFrame` y la `Series` coincide con el índice de la `Series` en las columnas del `DataFrame`, difundiéndose por las filas:

In [175]:
frame - series

b    d    e
Utah    0.0  0.0  0.0
Ohio    3.0  3.0  3.0
Texas   6.0  6.0  6.0
Oregon  9.0  9.0  9.0

Si no se encuentra un valor de índice ni en las columnas del `DataFrame` ni en el índice de la `Series`, los objetos se reindexarán para formar la unión:

In [176]:
series2 = pd.Series(np.arange(3), index=["b", "e", "f"])
series2


b    0
e    1
f    2
dtype: int64

In [177]:
frame + series2

b   d     e   f
Utah    0.0 NaN   3.0 NaN
Ohio    3.0 NaN   6.0 NaN
Texas   6.0 NaN   9.0 NaN
Oregon  9.0 NaN  12.0 NaN

Si, en cambio,se desea trabajar sobre las columnas, coincidiendo en las filas, debe utilizar uno de los métodos aritméticos y especificar que coincida sobre el índice. Por ejemplo:

In [178]:
series3 = frame["d"]
frame

b     d     e
Utah    0.0   1.0   2.0
Ohio    3.0   4.0   5.0
Texas   6.0   7.0   8.0
Oregon  9.0  10.0  11.0

In [179]:
series3

Utah       1.0
Ohio       4.0
Texas      7.0
Oregon    10.0
Name: d, dtype: float64

In [180]:
frame.sub(series3, axis="index")

b    d    e
Utah   -1.0  0.0  1.0
Ohio   -1.0  0.0  1.0
Texas  -1.0  0.0  1.0
Oregon -1.0  0.0  1.0

El eje que se pasa es el eje sobre el que se va a realizar la comparación. En este caso nos referimos a coincidir en el índice de fila del `DataFrame` `(axis="index")` y trabajará a través de las columnas.

### Aplicación y asignación de funciones
Los `ufuncs` de NumPy (métodos de array por elementos) también funcionan con objetos pandas:

In [181]:
frame = pd.DataFrame(np.random.standard_normal((4, 3)),
              columns=list("bde"),
              index=["Utah", "Ohio", "Texas", "Oregon"])
frame

b         d         e
Utah   -1.559417  1.466367  0.106946
Ohio    0.044929  1.180037 -0.506954
Texas  -0.522650 -1.243191 -0.487707
Oregon -0.635084  0.275049  0.097337

In [182]:
np.abs(frame)

b         d         e
Utah    1.559417  1.466367  0.106946
Ohio    0.044929  1.180037  0.506954
Texas   0.522650  1.243191  0.487707
Oregon  0.635084  0.275049  0.097337

Otra operación frecuente es aplicar una función en arrays unidimensionales a cada columna o fila. El método `apply` de DataFrame hace exactamente esto:

In [183]:
def f1(x):
    return x.max() - x.min()
frame.apply(f1)

b    1.604346
d    2.709558
e    0.613900
dtype: float64

Aquí la función `f`, que calcula la diferencia entre el máximo y el mínimo de una `Series`, se invoca una vez en cada columna de `frame`. El resultado es una `Series` que tiene como índice las columnas de `frame`.
Si se pasa `axis="columns"` al método `apply`, la función se invocará una vez por fila. Una forma útil de pensar en esto es como "aplicar a través de las columnas":

In [184]:
frame.apply(f1, axis="columns")

Utah      3.025784
Ohio      1.686991
Texas     0.755484
Oregon    0.910134
dtype: float64

Muchos de las funciones estadísticas de array más comunes (como suma y media) son métodos de DataFrame, por lo que no es necesario utilizar `apply`. No es necesario que la función pasada a `apply` devuelva un valor escalar; también puede devolver una Serie con múltiples valores:

In [185]:
frame

b         d         e
Utah   -1.559417  1.466367  0.106946
Ohio    0.044929  1.180037 -0.506954
Texas  -0.522650 -1.243191 -0.487707
Oregon -0.635084  0.275049  0.097337

In [186]:
def f2(x):
    return pd.Series([x.min(), x.max()], index=["min", "max"])

frame.apply(f2)

b         d         e
min -1.559417 -1.243191 -0.506954
max  0.044929  1.466367  0.106946

También se pueden utilizar funciones Python por elementos. Supongamos que desea calcular una cadena formateada a partir de cada valor de coma flotante de `frame` Puede hacerlo con `applymap`:

In [187]:
def my_format(x):
    return f"{x:.2f}"
frame.applymap(my_format)

C:\Users\juanj\AppData\Local\Temp\ipykernel_11568\1380983260.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  frame.applymap(my_format)


b      d      e
Utah    -1.56   1.47   0.11
Ohio     0.04   1.18  -0.51
Texas   -0.52  -1.24  -0.49
Oregon  -0.64   0.28   0.10

La razón del nombre `applymap` es que `Series` tiene un método `map` para aplicar una función `element-wise`:

In [188]:
frame["e"].map(my_format)

Utah       0.11
Ohio      -0.51
Texas     -0.49
Oregon     0.10
Name: e, dtype: object

### Clasificación y ordenación
Ordenar un conjunto de datos por algún criterio es otra importante operación incorporada. Para ordenar lexicográficamente por etiqueta de fila o columna, utilice el método `sort_index`, que devuelve un nuevo objeto ordenado:

In [189]:
obj = pd.Series(np.arange(4), index=["d", "a", "b", "c"])
obj

d    0
a    1
b    2
c    3
dtype: int64

In [190]:
obj.sort_index()

a    1
b    2
c    3
d    0
dtype: int64

Con un `DataFrame`,  se puede ordenar por índice en cualquiera de los ejes:

In [191]:
frame = pd.DataFrame(np.arange(8).reshape((2, 4)),
                     index=["three", "one"],
                     columns=["d", "a", "b", "c"])

frame

d  a  b  c
three  0  1  2  3
one    4  5  6  7

In [192]:
frame.sort_index()

d  a  b  c
one    4  5  6  7
three  0  1  2  3

In [193]:
frame.sort_index(axis="columns")

a  b  c  d
three  1  2  3  0
one    5  6  7  4

Por defecto, los datos se ordenan en orden ascendente, pero también pueden ordenarse en orden descendente:

In [194]:
frame.sort_index(axis="columns", ascending=False)

d  c  b  a
three  0  3  2  1
one    4  7  6  5

Para ordenar una `Serie` por sus valores, utilice su método `sort_values`:

In [195]:
obj = pd.Series([4, 7, -3, 2])
obj

0    4
1    7
2   -3
3    2
dtype: int64

In [196]:
obj.sort_values()

2   -3
3    2
0    4
1    7
dtype: int64

Los valores que faltan se ordenan por defecto al final de la serie:

In [197]:
obj = pd.Series([4, np.nan, 7, np.nan, -3, 2])
obj

0    4.0
1    NaN
2    7.0
3    NaN
4   -3.0
5    2.0
dtype: float64

In [198]:
obj.sort_values()

4   -3.0
5    2.0
0    4.0
2    7.0
1    NaN
3    NaN
dtype: float64

Los valores faltantes pueden ordenarse al principio utilizando la opción `na_position`:

In [199]:
obj.sort_values(na_position="first")

1    NaN
3    NaN
4   -3.0
5    2.0
0    4.0
2    7.0
dtype: float64

Al ordenar un DataFrame, puede utilizar los datos de una o varias columnas como claves de ordenación. Para ello, pase uno o más nombres de columna a `sort_values`:

In [200]:
frame = pd.DataFrame({"b": [4, 7, -3, 2], "a": [0, 1, 0, 1]})
frame

b  a
0  4  0
1  7  1
2 -3  0
3  2  1

In [201]:
frame.sort_values("b")

b  a
2 -3  0
3  2  1
0  4  0
1  7  1

Para ordenar por varias columnas, pase una lista de nombres:

In [202]:
frame.sort_values(["a", "b"])

b  a
2 -3  0
0  4  0
3  2  1
1  7  1

La clasificación `Ranking` asigna rangos desde uno hasta el número de puntos de datos válidos en una array, empezando por el valor más bajo. Los métodos `rank` para `Series` y `DataFrame` son el lugar donde buscar; por defecto, `rank` rompe los empates asignando a cada grupo el rango medio:

In [203]:
obj = pd.Series([7, -5, 7, 4, 2, 0, 4])
obj

0    7
1   -5
2    7
3    4
4    2
5    0
6    4
dtype: int64

In [204]:
obj.rank()

0    6.5
1    1.0
2    6.5
3    4.5
4    3.0
5    2.0
6    4.5
dtype: float64

También se pueden asignar rangos según el orden en que se observan en los datos:

In [205]:
obj.rank(method="first")

0    6.0
1    1.0
2    7.0
3    4.0
4    3.0
5    2.0
6    5.0
dtype: float64

Aquí, en lugar de utilizar el rango medio 6,5 para las entradas 0 y 2, se han fijado en 6 y 7 porque la etiqueta 0 precede a la etiqueta 2 en los datos.

También puedes clasificar en orden descendente:

obj

In [206]:
obj.rank(ascending=False)

0    1.5
1    7.0
2    1.5
3    3.5
4    5.0
5    6.0
6    3.5
dtype: float64

En ciencia de datos el método rank() puede usarse para identificar valores extremos, para identificar empates (duplicados), para ver la posibilidad de generar variables categóricas ordinales o que puede ser útil para aplicar modelos de aprendizaje automático que requieren entradas ordinales en lugar de valores numéricos continuos. 

Consulte la siguiente Tabla para ver una lista de los métodos de "desempate" (tie-breaking) disponibles.

<img src="tabla_6.png">

DataFrame puede calcular rangos sobre las filas o las columnas:

In [207]:
frame = pd.DataFrame({"b": [4.3, 7, -3, 2], "a": [0, 1, 0, 1],
                      "c": [-2, 5, 8, -2.5]})
frame

b  a    c
0  4.3  0 -2.0
1  7.0  1  5.0
2 -3.0  0  8.0
3  2.0  1 -2.5

In [208]:
frame.rank(axis="columns")

b    a    c
0  3.0  2.0  1.0
1  3.0  1.0  2.0
2  1.0  2.0  3.0
3  3.0  2.0  1.0

### Índices de ejes con etiquetas duplicadas
Hasta ahora casi todos los ejemplos que hemos visto tienen etiquetas de eje únicas (valores de índice). Aunque muchas funciones de pandas (como reindex) requieren que las etiquetas sean únicas, no es obligatorio. Consideremos una pequeña serie con índices duplicados:

In [ ]:
obj = pd.Series(np.arange(5), index=["a", "a", "b", "b", "c"])
obj

La propiedad `is_unique` del índice puede indicarle si sus etiquetas son únicas o no:

In [210]:
obj.index.is_unique

False

La selección de datos es una de las principales cosas que se comporta de forma diferente con los duplicados. La indexación de una etiqueta con varias entradas devuelve una Serie, mientras que las entradas únicas devuelven un valor escalar:

In [211]:
obj["a"]

a    0
a    1
dtype: int64

In [212]:
obj["c"]

np.int64(4)

Esto puede complicar su código, ya que el tipo de salida de la indexación puede variar en función de si una etiqueta se repite o no. La misma lógica se extiende a la indexación de filas (o columnas) en un DataFrame:

In [ ]:
df = pd.DataFrame(np.random.standard_normal((5, 3)),
                index=["a", "a", "b", "b", "c"])
df  

In [214]:
df.loc["b"]

0         1         2
b -0.749457 -0.826119 -0.737577
b -0.424646  0.478417 -0.032507

In [215]:
df.loc["c"]

0   -0.295179
1   -0.743866
2   -0.647579
Name: c, dtype: float64

### Resumir y calcular estadísticas descriptivas
Los objetos pandas están equipados con un conjunto de métodos matemáticos y estadísticos comunes. La mayoría de ellos entran en la categoría de reducciones o estadísticas de resumen, métodos que extraen un único valor (como la suma o la media) de una Serie, o una Serie de valores de las filas o columnas de un DataFrame. En comparación con los métodos similares que se encuentran en las matrices NumPy, tienen incorporado el manejo de los datos que faltan. Consideremos un pequeño DataFrame:

In [216]:
df = pd.DataFrame([[1.4, np.nan], [7.1, -4.5],
             [np.nan, np.nan], [0.75, -1.3]],
                index=["a", "b", "c", "d"],
                  columns=["one", "two"])
df

one  two
a  1.40  NaN
b  7.10 -4.5
c   NaN  NaN
d  0.75 -1.3

La llamada al método `sum` de DataFrame devuelve una `Serie` que contiene las sumas de las columnas:

In [217]:
df.sum()

one    9.25
two   -5.80
dtype: float64

Si se pasa `axis="columnas"` o `axis=1`, se suman las columnas:

In [218]:
df.sum(axis="columns")

a    1.40
b    2.60
c    0.00
d   -0.55
dtype: float64

Cuando una fila o columna entera contiene todos valores NA, la suma es 0, mientras que si algún valor no es NA, entonces el resultado es NA. Esto se puede desactivar con la opción `skipna`, en cuyo caso cualquier valor NA en una fila o columna nombra NA al resultado correspondiente:

In [219]:
df.sum(axis="index", skipna=False)

one   NaN
two   NaN
dtype: float64

In [220]:
df.sum(axis="columns", skipna=False)

a     NaN
b    2.60
c     NaN
d   -0.55
dtype: float64

Algunas agregaciones, como la media `mean`, requieren al menos un valor no-NA para producir un resultado de valor, así que aquí tenemos:

In [221]:
df.mean(axis="columns")

a    1.400
b    1.300
c      NaN
d   -0.275
dtype: float64

Véase la siguiente tabla una lista de opciones habituales para cada método de reducción:

<img src="tabla_6.png">

Algunos métodos, como `idxmin` e `idxmax`, devuelven estadísticas indirectas, como el valor del índice donde se alcanzan los valores mínimo o máximo:

In [222]:
df

one  two
a  1.40  NaN
b  7.10 -4.5
c   NaN  NaN
d  0.75 -1.3

In [223]:
df.idxmax()

one    b
two    d
dtype: object

Otros métodos son las acumulaciones:

In [224]:
df.cumsum()

one  two
a  1.40  NaN
b  8.50 -4.5
c   NaN  NaN
d  9.25 -5.8

In [225]:
df

one  two
a  1.40  NaN
b  7.10 -4.5
c   NaN  NaN
d  0.75 -1.3

Algunos métodos no son ni reducciones ni acumulaciones. `describe` es un ejemplo de ello, ya que produce múltiples estadísticas de resumen de una sola vez:

In [226]:
df.describe()

one       two
count  3.000000  2.000000
mean   3.083333 -2.900000
std    3.493685  2.262742
min    0.750000 -4.500000
25%    1.075000 -3.700000
50%    1.400000 -2.900000
75%    4.250000 -2.100000
max    7.100000 -1.300000

En datos no numéricos, describe produce estadísticas de resumen alternativas:

In [227]:
obj = pd.Series(["a", "a", "b", "c"] * 4)
obj

0     a
1     a
2     b
3     c
4     a
5     a
6     b
7     c
8     a
9     a
10    b
11    c
12    a
13    a
14    b
15    c
dtype: object

In [228]:
obj.describe()

count     16
unique     3
top        a
freq       8
dtype: object

Véase la siguiente tabla para una lista completa de estadísticas de síntesis y métodos relacionados.

<img src="tabla_8.png">

## Correlación y covarianza

### Covarianza

La covarianza nos dá la covariación conjunta de dos variables. Es decir como cambia una variable respecto a otra.  
Si la covarianza es positiva nos dará la información de que a valores altos de una de las variable (X) hay una mayor tendencia a encontrar valores altos de la otra variable (Y) y a valores bajos de una de las variable ,correspondientemente valores bajos de la otra. 

![eq](coveq.png)

La covarianza se suele representar por $S_{sy}$ y suele obtenerse dos valores:  

Si $S_{xy}$ es positivo la covarianza es directa. A medida que aumenta X tambien aumenta Y.  

Si $S_{xy}$ es negativo la covarianza es inversa. A medida que aumenta X la Y disminuye.  
Si $S_{xy}$ es cero no hay covarianza

![Covarianzas](covarianza.png)

![Covarianza_0](cov_0.png)

La covarianza presenta como inconveniente, el hecho de que su valor depende de la escala elegida para los ejes. 
Es decir, la covarianza variará si expresamos la altura en metros o en centímetros. También variará si el dinero lo expresamos en euros o en dólares.

La interpretación de la covarianza es con el signo, los números que da por si solos no ofrecen ninguna interpretación salvo que sean muy cercanos a cero.

### Correlación

La correlación (Pearson) mide tanto la dirección como la fuerza de la relación lineal entre dos variables. Los valores de la correlación están normalizados entre -1 y 1. Una correlación de 1 indica una relación lineal perfecta positiva, -1 indica una relación lineal perfecta negativa, y 0 indica que no hay una relación lineal.

![correlacion](cor_pear.png)

![graf](graf.png)

Algunos estadísticos de resumen, como la correlación y la covarianza, se calculan a partir de pares de argumentos. Consideremos algunos DataFrames de precios y volúmenes de acciones obtenidos originalmente de Yahoo! Finance:

In [256]:
price = pd.read_pickle("yahoo_price.pkl")
volume = pd.read_pickle("yahoo_volume.pkl")

- `read_pickle()`se utiliza para leer un objeto almacenado en un archivo en formato pickle. Pickle es un formato binario en Python que permite serializar y deserializar estructuras de datos de Python, lo que significa que puedes guardar un objeto en un archivo y luego leer ese archivo para recuperar el objeto en su forma original.
    - Serialización: Es el proceso de convertir un objeto en memoria en una secuencia de bytes que puede ser almacenada en un archivo o transmitida a través de una red.
    - Deserialización: Es el proceso inverso, donde la secuencia de bytes se convierte de nuevo en el objeto original en memoria.

In [258]:
price.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 1714 entries, 2010-01-04 to 2016-10-21
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   AAPL    1714 non-null   float64
 1   GOOG    1714 non-null   float64
 2   IBM     1714 non-null   float64
 3   MSFT    1714 non-null   float64
dtypes: float64(4)
memory usage: 67.0 KB


In [259]:
price.head()

AAPL        GOOG         IBM       MSFT
Date                                                    
2010-01-04  27.990226  313.062468  113.304536  25.884104
2010-01-05  28.038618  311.683844  111.935822  25.892466
2010-01-06  27.592626  303.826685  111.208683  25.733566
2010-01-07  27.541619  296.753749  110.823732  25.465944
2010-01-08  27.724725  300.709808  111.935822  25.641571

La tabla anterior podría leerse de esta forma:  

El 4 de enero de 2010, el precio ajustado de cierre en dólares de una acción de Apple fue de 27.99, el de Google fue de 313.06, el de IBM fue de 113.30 y el de Microsoft fue de 25.88.

In [286]:
volume.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 1714 entries, 2010-01-04 to 2016-10-21
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   AAPL    1714 non-null   int64
 1   GOOG    1714 non-null   int64
 2   IBM     1714 non-null   int64
 3   MSFT    1714 non-null   int64
dtypes: int64(4)
memory usage: 67.0 KB


In [260]:
volume.head()

AAPL      GOOG      IBM      MSFT
Date                                              
2010-01-04  123432400   3927000  6155300  38409100
2010-01-05  150476200   6031900  6841400  49749600
2010-01-06  138040000   7987100  5605300  58182400
2010-01-07  119282800  12876600  5840600  50559700
2010-01-08  111902700   9483900  4197200  51197400

La tabla representa los volúmenes de negociación diarios para las acciones de cuatro grandes compañías tecnológicas: Apple (AAPL), Google (GOOG), IBM y Microsoft (MSFT). El volumen de negociación se refiere al número total de acciones que se intercambiaron (compraron y vendieron) durante el día.

Volviendo a la dataframe price, por ejemplo, se podría calcular las variaciones porcentuales de los precios:

In [287]:
returns = price.pct_change()

Siendo el cambio porcentual:  

$pct\_change = \frac{valor\ actual - valor\ anterior}{valor\ anterior}$

In [288]:
returns

AAPL      GOOG       IBM      MSFT
Date                                              
2010-01-04       NaN       NaN       NaN       NaN
2010-01-05  0.001729 -0.004404 -0.012080  0.000323
2010-01-06 -0.015906 -0.025209 -0.006496 -0.006137
2010-01-07 -0.001849 -0.023280 -0.003462 -0.010400
2010-01-08  0.006648  0.013331  0.010035  0.006897
...              ...       ...       ...       ...
2016-10-17 -0.000680  0.001837  0.002072 -0.003483
2016-10-18 -0.000681  0.019616 -0.026168  0.007690
2016-10-19 -0.002979  0.007846  0.003583 -0.002255
2016-10-20 -0.000512 -0.005652  0.001719 -0.004867
2016-10-21 -0.003930  0.003011 -0.012474  0.042096

[1714 rows x 4 columns]

El primer valor en cada columna será `NaN` (Not a Number) porque no hay un valor anterior con el cual calcular el cambio.  
Por ejemplo, para la fecha 2010-01-05:  

- `AAPL`: **0.001729** indica que el precio de Apple aumentó en un 0.1729% respecto al día anterior.  

- `GOOG`: **-0.004404** indica que el precio de Google disminuyó en un 0.4404% respecto al día anterior.
- `IBM`: **-0.012080** indica que el precio de IBM disminuyó en un 1.208% respecto al día anterior.
- `MSFT`: **0.000323** indica que el precio de Microsoft aumentó en un 0.0323% respecto al día anterior.

In [263]:
returns.tail()

AAPL      GOOG       IBM      MSFT
Date                                              
2016-10-17 -0.000680  0.001837  0.002072 -0.003483
2016-10-18 -0.000681  0.019616 -0.026168  0.007690
2016-10-19 -0.002979  0.007846  0.003583 -0.002255
2016-10-20 -0.000512 -0.005652  0.001719 -0.004867
2016-10-21 -0.003930  0.003011 -0.012474  0.042096

El método `corr` de Series calcula la correlación de los valores superpuestos, no-NA, alineados-por-índice en dos Series. Por su parte, `cov` calcula la covarianza:

In [264]:
returns["MSFT"].corr(returns["IBM"])

np.float64(0.4997636114415114)

In [265]:
returns["MSFT"].cov(returns["IBM"])

np.float64(8.870655479703545e-05)

Los métodos `corr` y `cov` de un DataFrame, por otro lado, devuelven una matriz de correlación o covarianza completa como un DataFrame, respectivamente:

In [266]:
returns.corr()

AAPL      GOOG       IBM      MSFT
AAPL  1.000000  0.407919  0.386817  0.389695
GOOG  0.407919  1.000000  0.405099  0.465919
IBM   0.386817  0.405099  1.000000  0.499764
MSFT  0.389695  0.465919  0.499764  1.000000

In [267]:
returns.cov()

AAPL      GOOG       IBM      MSFT
AAPL  0.000277  0.000107  0.000078  0.000095
GOOG  0.000107  0.000251  0.000078  0.000108
IBM   0.000078  0.000078  0.000146  0.000089
MSFT  0.000095  0.000108  0.000089  0.000215

Utilizando el método `corrwith` de DataFrame, puede calcular correlaciones por pares entre las columnas o filas de un DataFrame con otra Serie o DataFrame. Al pasar una Serie se devuelve una Serie con el valor de correlación calculado para cada columna:

In [268]:
returns.corrwith(returns["IBM"])

AAPL    0.386817
GOOG    0.405099
IBM     1.000000
MSFT    0.499764
dtype: float64

Al pasar un DataFrame se calculan las correlaciones de los nombres de columna coincidentes. Aquí, se calculan las correlaciones de los cambios porcentuales con el volumen:

In [269]:
returns.corrwith(volume)

AAPL   -0.075565
GOOG   -0.007067
IBM    -0.204849
MSFT   -0.092950
dtype: float64

Si se pasa `axis="columns"`, se hace fila por fila. En todos los casos, los puntos de datos se alinean por etiqueta antes de calcular la correlación.

Recordemos que:

![correlaciones](cor_val.webp)

### Valores únicos, recuento de valores y afiliación

Otra clase de métodos relacionados extrae información sobre los valores contenidos en una Serie unidimensional. Para ilustrarlos, considere este ejemplo:

In [271]:
obj = pd.Series(["c", "a", "d", "a", "a", "b", "b", "c", "c"])
obj

0    c
1    a
2    d
3    a
4    a
5    b
6    b
7    c
8    c
dtype: object

La primera función es `unique`, que nos da un array de los valores únicos de una Serie:

In [272]:
uniques = obj.unique()
uniques

array(['c', 'a', 'd', 'b'], dtype=object)

Los valores únicos no se devuelven necesariamente en el orden en que aparecen por primera vez, y no en orden ordenado, pero podrían ordenarse a posteriori si fuera necesario (`uniques.sort()`). Por otro lado, `value_counts` calcula una serie que contiene las frecuencias de los valores:

In [273]:
obj.value_counts()

c    3
a    3
b    2
d    1
Name: count, dtype: int64

La serie se ordena por valor en orden descendente por conveniencia. `value_counts` también está disponible como un método pandas de nivel superior que se puede utilizar con arrays NumPy u otras secuencias de Python:

In [274]:
 pd.value_counts(obj.to_numpy(), sort=False)

C:\Users\juanj\AppData\Local\Temp\ipykernel_11568\2238870943.py:1: FutureWarning: pandas.value_counts is deprecated and will be removed in a future version. Use pd.Series(obj).value_counts() instead.
  pd.value_counts(obj.to_numpy(), sort=False)


c    3
a    3
d    1
b    2
Name: count, dtype: int64

`isin` realiza una comprobación de pertenencia a un conjunto vectorizado (sin bucles) y puede ser útil para filtrar un conjunto de datos a un subconjunto de valores en una Serie o columna en un DataFrame:

In [275]:
obj

0    c
1    a
2    d
3    a
4    a
5    b
6    b
7    c
8    c
dtype: object

In [276]:
mask = obj.isin(["b", "c"])
mask

0     True
1    False
2    False
3    False
4    False
5     True
6     True
7     True
8     True
dtype: bool

In [277]:
obj[mask]

0    c
5    b
6    b
7    c
8    c
dtype: object

Relacionado con `isin` está el método `Index.get_indexer`, que es útil para mapear y alinear un conjunto de valores con los índices de otro conjunto de valores. En otras palabras, te permite encontrar las posiciones de los elementos de un array (o lista) en otro array (o lista).

In [316]:
to_match = pd.Series(["c", "a", "b", "b", "c", "a"])
# Estos son los valores que queremos mapear a sus índices en unique_vals.
to_match

0    c
1    a
2    b
3    b
4    c
5    a
dtype: object

In [315]:
unique_vals = pd.Series(["c", "b", "a"])
# Estos valores son los de referencia para encontrar los índices de los valores en to_match.
unique_vals

0    c
1    b
2    a
dtype: object

In [ ]:
indices = pd.Index(unique_vals).get_indexer(to_match)
# pd.Index(unique_vals): Convierte unique_vals en un objeto Index de pandas.
# .get_indexer(to_match): Encuentra los índices de los valores en to_match dentro de unique_vals.
indices

El método get_indexer devuelve un array de índices (únicos) correspondientes a las posiciones de los elementos de to_match en unique_vals. Si un elemento de to_match no se encuentra en unique_vals, get_indexer devolverá -1 para ese elemento.

- to_match[0] ("c"):  La posición de "c" en unique_vals es 0.
- to_match[1] ("a"): La posición de "a" en unique_vals es 2.
- to_match[2] ("b"): La posición de "b" en unique_vals es 1.
- to_match[3] ("b"): La posición de "b" en unique_vals es 1.
- to_match[4] ("c"): La posición de "c" en unique_vals es 0.
- to_match[5] ("a"): La posición de "a" en unique_vals es 2.

Esto es útil para reindexar, mapear o alinear datos entre diferentes Series o DataFrames en pandas de manera eficiente.

Otro ejemplo:

Supongamos que tienes dos arrays:

1 - Array de referencia (valores únicos):

In [298]:
referencia = pd.Index(['a', 'b', 'c', 'd'])
referencia

Index(['a', 'b', 'c', 'd'], dtype='object')

2 - Array de búsqueda (valores que pueden repetirse y no necesariamente todos están en el array de referencia):

In [297]:
busqueda = pd.Series(['a', 'c', 'e', 'c'])
busqueda

0    a
1    c
2    e
3    c
dtype: object

Usando `get_indexer`:

In [299]:
import pandas as pd

# Array de referencia
referencia = pd.Index(['a', 'b', 'c', 'd'])

# Array de búsqueda
busqueda = pd.Series(['a', 'c', 'e', 'c'])

# Obtener los índices
indices = referencia.get_indexer(busqueda)

indices

array([ 0,  2, -1,  2])

- 0: El primer valor de busqueda ('a') se encuentra en la posición 0 del array referencia.
- 2: El segundo valor de busqueda ('c') se encuentra en la posición 2 del array referencia.
- -1: El tercer valor de busqueda ('e') no se encuentra en el array referencia, por lo que devuelve -1.
- 2: El cuarto valor de busqueda ('c') nuevamente se encuentra en la posición 2 del array referencia.

**Utilidad**  
- Alineación: Puedes alinear datos entre diferentes estructuras, como cuando tienes un DataFrame y necesitas reindexarlo basándote en otro conjunto de índices.
- Búsqueda eficiente: Permite buscar de manera eficiente las posiciones de múltiples elementos en un array de referencia.

**Recuentos**

En algunos casos, es posible que desee calcular un histograma en varias columnas relacionadas en un DataFrame. He aquí un ejemplo:

In [308]:
data = pd.DataFrame({"Qu1": [1, 3, 4, 3, 4],
                     "Qu2": [2, 3, 1, 2, 3],
                     "Qu3": [1, 5, 2, 4, 4]})
data

Qu1  Qu2  Qu3
0    1    2    1
1    3    3    5
2    4    1    2
3    3    2    4
4    4    3    4

Podemos calcular los recuentos de valores para una sola columna, de la siguiente manera:

In [310]:
data["Qu1"].value_counts()
# Contar la frecuencia de los valores en la columna "Qu1"

Qu1
3    2
4    2
1    1
Name: count, dtype: int64

 - El método `value_counts()` cuenta la frecuencia de cada valor único en la columna "Qu1". Esto produce una Serie con los valores únicos de "Qu1" como índices y sus frecuencias como valores.
 - El valor 3 aparece 2 veces, el valor 4 aparece 2 veces y el valor 1 aparece 1 vez en la columna "Qu1".

In [311]:
data["Qu1"].value_counts().sort_index()

Qu1
1    1
3    2
4    2
Name: count, dtype: int64

- El método `sort_index()` ordena la Serie resultante de `value_counts()` por los índices (los valores únicos de "Qu1") en orden ascendente.
    - 1 aparece 1 vez: Indica que el valor 1 aparece una vez en la columna "Qu1".
    - 3 aparece 2 veces: Indica que el valor 3 aparece dos veces en la columna "Qu1".
    - 4 aparece 2 veces: Indica que el valor 4 aparece dos veces en la columna "Qu1".

### Este proceso es útil para entender la distribución de los datos en una columna específica, facilitando el análisis de frecuencias y patrones dentro de los datos

Para calcular lo anterior pero para todas las columnas, pase `pandas.value_counts` al método `apply` del DataFrame:

In [282]:
result = data.apply(pd.value_counts).fillna(0)
result

C:\Users\juanj\AppData\Local\Temp\ipykernel_11568\1382616601.py:1: FutureWarning: pandas.value_counts is deprecated and will be removed in a future version. Use pd.Series(obj).value_counts() instead.
  result = data.apply(pd.value_counts).fillna(0)


Qu1  Qu2  Qu3
1  1.0  1.0  1.0
2  0.0  2.0  1.0
3  2.0  2.0  0.0
4  2.0  0.0  2.0
5  0.0  0.0  1.0

Aquí, las etiquetas de fila en el resultado son los valores distintos que aparecen en todas las columnas. Los valores son los recuentos respectivos de estos valores en cada columna.

También existe un método `DataFrame.value_counts`, pero éste calcula los recuentos considerando cada fila del DataFrame como una tupla para determinar el número de ocurrencias de cada fila distinta:

In [312]:
data = pd.DataFrame({"a": [1, 1, 1, 2, 2], "b": [0, 0, 1, 0, 0]})
data

a  b
0  1  0
1  1  0
2  1  1
3  2  0
4  2  0

In [313]:
data.value_counts()

a  b
1  0    2
2  0    2
1  1    1
Name: count, dtype: int64